### PR TITLE
Add a battery sensor to battery powered devices

### PR DESCRIPTION
**Pull request by stefanbode**
This should work without modifications in HA and enable the battery magic in HA.

Tasmota Command:
`batterypercentage 45` to set something. `restart 1` required if not enabled before

`batterypercentage 101` to disable battery., `restart 1` required

### CODE REVIEW

File: hatasmota/const.py [hunk 1]
```diff
@@ -36,6 +36,8 @@
 CONF_OPTIONS: Final = "so"
 CONF_PREFIX: Final = "tp"
 CONF_SENSOR: Final = "sn"
+CONF_BATTERY: Final = "bat"
+CONF_DEEPSLEEP: Final = "dslp"
```
Comment (emontnemery) [inline hatasmota/const.py:40]:
```suggestion
```

File: hatasmota/const.py [hunk 2]
```diff
@@ -92,6 +94,10 @@
 RSLT_STATE: Final = "STATE"
 RSLT_TRIG: Final = "TRIG"
 
+DEEPSLEEP_REPORTTIME: Final = "Time"
+DEEPSLEEP_SLEEPTIME: Final = "DeepSleep"
+DEEPSLEEP_WAKEUPTIME: Final = "Wakeup"
+
```
Comment (emontnemery) [inline hatasmota/const.py:100]:
```suggestion
```

File: hatasmota/const.py [hunk 3]
```diff
@@ -161,6 +167,8 @@
 SENSOR_STATUS_SSID: Final = "status_ssid"
 SENSOR_STATUS_VERSION: Final = "status_version"
 
+SENSOR_STATUS_DEEPSLEEP: Final = "status_deepdleep"
```
Comment (emontnemery) [inline hatasmota/const.py:170]: Let's add the battery one too. The `status_deepsleep` can be held off to the next PR though.
```suggestion
SENSOR_STATUS_BATTERY_PERCENTAGE: Final = "status_battery_percentage"
```

File: hatasmota/device_status.py [hunk 4]
```diff
@@ -38,6 +38,7 @@
 #  "Module or Template":"Generic",             <NONE>
 #  "RestartReason":"Software/System restart",  stat/STATUS1:"StatusPRM"."RestartReason"
 #  "Uptime":"1T17:04:28",                      stat/STATUS11:"StatusSTS"."Uptime"; tele/STATE:"Uptime"
+#  "BatteryPercentage":60,                     // optional since 13.0.0.3 if battery defined
```
Comment (emontnemery) [inline hatasmota/device_status.py:41]:
```suggestion
#  "BatteryPercentage":60,                     tele/STATE: "BatteryPercentage"
```

File: hatasmota/discovery.py [hunk 5]
```diff
@@ -29,6 +29,8 @@
     CONF_PREFIX,
     CONF_RELAY,
     CONF_SENSOR,
+    CONF_BATTERY,
+    CONF_DEEPSLEEP,
```
Comment (emontnemery) [inline hatasmota/discovery.py:33]:
```suggestion
```

File: hatasmota/discovery.py [hunk 6]
```diff
@@ -66,6 +68,7 @@
 from .mqtt import ReceiveMessage, TasmotaMQTTClient
 from .relay import TasmotaRelay, TasmotaRelayConfig
 from .sensor import TasmotaBaseSensorConfig, TasmotaSensor, get_sensor_entities
+# from .deepsleep import TasmotaDeepSleep
```
Comment (emontnemery) [inline hatasmota/discovery.py:71]:
```suggestion
```

File: hatasmota/discovery.py [hunk 7]
```diff
@@ -122,9 +125,11 @@
         CONF_OPTIONS: TASMOTA_OPTIONS_SCHEMA,
         CONF_PREFIX: vol.All(cv.ensure_list, [cv.string]),
         CONF_STATE: vol.All(cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_BATTERY, default=0): cv.positive_int, # Added in Tasmota 13.0.0.3
+        vol.Optional(CONF_DEEPSLEEP, default=0): cv.positive_int, # Added in Tasmota 13.0.0.3
         vol.Optional(CONF_SHUTTER_OPTIONS, default=[]): vol.All(
             cv.ensure_list, [cv.positive_int]
-        ),  # Added in Tasmota 9.2
+        ),
```
Comment (emontnemery) [inline hatasmota/discovery.py:132]: Why was this comment removed?

File: hatasmota/discovery.py [hunk 8]
```diff
@@ -122,9 +125,11 @@
         CONF_OPTIONS: TASMOTA_OPTIONS_SCHEMA,
         CONF_PREFIX: vol.All(cv.ensure_list, [cv.string]),
         CONF_STATE: vol.All(cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_BATTERY, default=0): cv.positive_int, # Added in Tasmota 13.0.0.3
+        vol.Optional(CONF_DEEPSLEEP, default=0): cv.positive_int, # Added in Tasmota 13.0.0.3
```
Comment (emontnemery) [inline hatasmota/discovery.py:129]:
```suggestion
```

File: hatasmota/status_sensor.py [hunk 9]
```diff
@@ -13,12 +13,14 @@
...
     CONF_IP,
     CONF_MAC,
+    CONF_BATTERY,
     PERCENTAGE,
     SENSOR_STATUS_IP,
     SENSOR_STATUS_LAST_RESTART_TIME,
     SENSOR_STATUS_LINK_COUNT,
     SENSOR_STATUS_MQTT_COUNT,
     SENSOR_STATUS_RESTART_REASON,
+    SENSOR_BATTERY,
```
Comment (emontnemery) [inline hatasmota/status_sensor.py:23]:
```suggestion
    SENSOR_BATTERY,
    SENSOR_STATUS_BATTERY_PERCENTAGE,
```

File: hatasmota/status_sensor.py [hunk 10]
```diff
@@ -74,6 +77,7 @@
     SENSOR_STATUS_LINK_COUNT: "WiFi Connect Count",
     SENSOR_STATUS_MQTT_COUNT: "MQTT Connect Count",
     SENSOR_STATUS_RESTART_REASON: "Restart Reason",
+    SENSOR_BATTERY: "Battery Level",
```
Comment (emontnemery) [inline hatasmota/status_sensor.py:80]: We don't need a name for this, Home Assistant automatically sets the name of battery sensors
```suggestion
```

File: hatasmota/status_sensor.py [hunk 11]
```diff
@@ -89,6 +93,7 @@
 STATE_PATHS: dict[str, list[str | int]] = {
     SENSOR_STATUS_LINK_COUNT: ["Wifi", "LinkCount"],
     SENSOR_STATUS_MQTT_COUNT: ["MqttCount"],
+    SENSOR_BATTERY: ["BatteryPercentage"],
```
Comment (emontnemery) [inline hatasmota/status_sensor.py:96]:
```suggestion
    SENSOR_STATUS_BATTERY_PERCENTAGE: ["BatteryPercentage"],
```

File: hatasmota/status_sensor.py [hunk 12]
```diff
@@ -121,6 +126,7 @@
     SENSOR_STATUS_LINK_COUNT: SENSOR_STATUS_LINK_COUNT,
     SENSOR_STATUS_MQTT_COUNT: SENSOR_STATUS_MQTT_COUNT,
     SENSOR_STATUS_RESTART_REASON: SENSOR_STATUS_RESTART_REASON,
+    SENSOR_BATTERY: SENSOR_BATTERY,
```
Comment (emontnemery) [inline hatasmota/status_sensor.py:129]:
```suggestion
    SENSOR_STATUS_BATTERY_PERCENTAGE: SENSOR_BATTERY,
```

File: hatasmota/status_sensor.py [hunk 13]
```diff
@@ -133,6 +139,7 @@
     SENSOR_STATUS_LINK_COUNT: None,
     SENSOR_STATUS_MQTT_COUNT: None,
     SENSOR_STATUS_RESTART_REASON: None,
+    SENSOR_BATTERY: PERCENTAGE,
```
Comment (emontnemery) [inline hatasmota/status_sensor.py:142]:
```suggestion
    SENSOR_STATUS_BATTERY_PERCENTAGE: PERCENTAGE,
```

File: hatasmota/status_sensor.py [hunk 14]
```diff
@@ -155,6 +162,9 @@ def from_discovery_message(
         cls, config: dict, platform: str
     ) -> list[TasmotaStatusSensorConfig]:
         """Instantiate from discovery message."""
+        sensor_types = list(SENSORS)
+        if config[CONF_BATTERY] == 1:
+            sensor_types.append(SENSOR_BATTERY)
```
Comment (emontnemery) [inline hatasmota/status_sensor.py:167]:
```suggestion
            sensor_types.append(SENSOR_STATUS_BATTERY_PERCENTAGE)
```

**Review by emontnemery — CHANGES_REQUESTED** [review]
This is great!

I added some suggested changes though:
- Let's remove the deep sleep related changes from this PR.
- I also think we should better separate the battery percentage sensor from a "normal" battery sensor, hence a separate constant for that

File: hatasmota/status_sensor.py (hunk 10 above)
Comment (emontnemery) [reply hatasmota/status_sensor.py:80]: Never mind this comment, the Tasmota integration has not yet opted in to this feature

File: hatasmota/discovery.py (hunk 7 above)
Comment (stefanbode) [reply hatasmota/discovery.py:130]: sorry. readded

File: hatasmota/discovery.py [hunk 15]
```diff
@@ -122,6 +123,7 @@
         CONF_OPTIONS: TASMOTA_OPTIONS_SCHEMA,
         CONF_PREFIX: vol.All(cv.ensure_list, [cv.string]),
         CONF_STATE: vol.All(cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_BATTERY, default=0): cv.positive_int, # Added in Tasmota 13.0.0.3
```
Comment (emontnemery) [inline hatasmota/discovery.py:126]:
```suggestion
        vol.Optional(
            CONF_BATTERY, default=0
        ): cv.positive_int,  # Added in Tasmota 13.0.0.3
```

**Review by emontnemery — APPROVED** [review]
Thanks a lot, @stefanbode 👍

**Pull request merged**